### PR TITLE
pkg/dns/azure: implement support for azure

### PR DESCRIPTION
**Pull request by serbrech**
The azure dns zones do not support ALIAS records.
We use A records instead.

### CODE REVIEW

**Comment by ironcladlou** [comment]
/hold

Thanks for taking a stab at this! Before you invest much more time in the patch, you should know we're actively working to replace the internal DNS manager with an [external-dns](https://github.com/kubernetes-incubator/external-dns) controller implementation.

Do you have any reason to suspect we'll encounter problems with the external-dns Azure provider?

**Comment by ironcladlou** [comment]
I've been reviewing this PR, and it's looking really good. Thank you!

In the background I'm working on figuring out how we can validate this outside of CI. Right now it's looking like we'd want to try and accept this when master re-opens for the 4.2 release, even in advance of our external-dns plans.

File: pkg/dns/azure/dns.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	return m.client.Put(
+		context.TODO(),
+		*targetZone,
+		client.ARecord{
+			Address: record.ARecord.Address,
+			Name:    ARecordName,
+		})
+}
+
+func (m *manager) Delete(record *dns.Record) error {
```
Comment (ironcladlou) [inline pkg/dns/azure/dns.go]: I take it this function is still a WIP (as it doesn't seem to actually delete, hence the unit test failure)?

**Comment by serbrech** [comment]
Thanks for the quick feedback!
There are a couple things I need to validate before I'd remove [WIP], and I didn't run it on a cluster yet.

**Q:** Why is there a cache for AWS? is that something we'd need to do as well?

I think it makes sense to contribute to `external-dns` as it's doing pretty much the same thing.
We've used it successfully up to a very high scale, it's solid. 

From browsing their code, it finds the Azure DNSZone [by resource group](https://github.com/kubernetes-incubator/external-dns/blob/ad68fb8daf3794475d152f57856225fd831009d8/provider/azure.go#L228). 
In the openshift installer, we require the public dns zone to be existing in the subscription prior to deployment, so it is typically in a different RG.
That means it would require some changes, either in external-dns, or in openshift installer.

Meanwhile, this PR, once completed, should be enough to move forward until you make the switch.

File: pkg/dns/azure/dns.go (hunk 1 above)
Comment (serbrech) [reply pkg/dns/azure/dns.go]: that was a miss when restructuring my commits. thanks :)

File: pkg/dns/azure/client/auth.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		return nil, err
+	}
+
+	token, err := adal.NewServicePrincipalToken(
+		*oauthConfig, config.ClientID, config.ClientSecret, env.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	a = autorest.NewBearerAuthorizer(token)
+	return a, err
```
Comment (abhinavdahiya) [inline pkg/dns/azure/client/auth.go]: nit: why not `return autorest.NewBearerAuthorizer(token), err`

**Comment by ironcladlou** [comment]
@serbrech 
>Q: Why is there a cache for AWS? is that something we'd need to do as well?

It's a cheap/quick local optimization to drastically reduce chattiness with the Route53 API. I won't be surprised if something similar is needed for Azure in practice — will think on it more.

>I think it makes sense to contribute to external-dns as it's doing pretty much the same thing.
We've used it successfully up to a very high scale, it's solid.

Thanks for the feedback. I'm encouraged to hear that!

>From browsing their code, it finds the Azure DNSZone by resource group.
In the openshift installer, we require the public dns zone to be existing in the subscription prior to deployment, so it is typically in a different RG.
That means it would require some changes, either in external-dns, or in openshift installer.

Not entirely sure I understand. Are proposing some change to the public zone type in the [OpenShift cluster config](https://github.com/openshift/api/blob/master/config/v1/types_dns.go#L35) API? (That's  currently the only DNS configuration recognized by the ingress operator.)

**Comment by serbrech** [comment]
>Not entirely sure I understand. Are proposing some change to the public zone type in the OpenShift cluster config API? (That's currently the only DNS configuration recognized by the ingress operator.)

There is 2 dns zones. private, and public.
the ingress operator is [inserting records in both](https://github.com/openshift/cluster-ingress-operator/blob/1cefef27939c65cb58c70e18d9b70fde23372329/pkg/operator/controller/controller_dns.go#L71-L76)

That is not supported by the external-dns.

File: pkg/dns/azure/dns.go (hunk 1 above)
Comment (serbrech) [reply pkg/dns/azure/dns.go]: fixed, added logging to match aws `log.Info("deleted/upserted DNS record", "record", record)`

**Comment by danehans** [comment]
> That is not supported by external-dns.

The ExternalDNS [faq](https://github.com/kubernetes-incubator/external-dns/blob/master/docs/faq.md#running-an-internal-and-external-dns-service) covers this limitation in more detail.

**Comment by serbrech** [comment]
logs : 
```
% KUBECONFIG=~/tmp/foo5/auth/kubeconfig RELEASE_VERSION=4.2.0-0.okd-2019-05-18-050412 IMAGE=registry.svc.ci.openshift.org/origin/4.2-2019-05-18-050412@sha256:d147b0db308cd33aae85c0147cbc7c0883148e8f197f24e579a6a78d16500e6b WATCH_NAMESPACE=openshift-ingress-operator ./ingress-operator
2019-05-19T10:39:26.364-0700    INFO    operator        log/log.go:26   started zapr logger
2019-05-19T10:39:32.239-0700    INFO    operator.entrypoint     ingress-operator/main.go:154    using azure creds from secret   {"namespace": "openshift-ingress-operator", "name": "cloud-credentials"}
2019-05-19T10:39:48.279-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "operator-controller", "source": "kind source: /, Kind="}
2019-05-19T10:39:52.971-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "operator-controller", "source": "informer source: 0xc0000eea20"}
2019-05-19T10:39:52.972-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "operator-controller", "source": "informer source: 0xc0000eeab0"}
2019-05-19T10:39:52.972-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "certificate-controller", "source": "kind source: /, Kind="}
2019-05-19T10:39:52.972-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "certificate-publisher-controller", "source": "informer source: 0xc0000eee10"}
2019-05-19T10:39:52.972-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:121    Starting EventSource    {"controller": "certificate-publisher-controller", "source": "kind source: /, Kind="}
2019-05-19T10:39:52.972-0700    INFO    operator.init   operator/operator.go:187        waiting for cache to sync
2019-05-19T10:39:53.138-0700    INFO    operator.init   operator/operator.go:126        queueing ingress        {"name": "default", "related": "/api/v1/namespaces/openshift-ingress/services/router-internal-default"}
2019-05-19T10:39:53.138-0700    INFO    operator.init   operator/operator.go:126        queueing ingress        {"name": "default", "related": "/api/v1/namespaces/openshift-ingress/services/router-default"}
2019-05-19T10:39:53.216-0700    INFO    operator.init   operator/operator.go:126        queueing ingress        {"name": "default", "related": "/apis/apps/v1/namespaces/openshift-ingress/deployments/router-default"}
2019-05-19T10:39:54.773-0700    INFO    operator.init   operator/operator.go:191        cache synced
2019-05-19T10:39:54.875-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:134    Starting Controller     {"controller": "certificate-publisher-controller"}
2019-05-19T10:39:54.875-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:134    Starting Controller     {"controller": "operator-controller"}
2019-05-19T10:39:54.875-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:134    Starting Controller     {"controller": "certificate-controller"}
2019-05-19T10:39:54.975-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:154    Starting workers        {"controller": "certificate-publisher-controller", "worker count": 1}
2019-05-19T10:39:54.976-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:154    Starting workers        {"controller": "operator-controller", "worker count": 1}
2019-05-19T10:39:54.976-0700    INFO    operator.init.controller-runtime.controller     controller/controller.go:154    Starting workers        {"controller": "certificate-controller", "worker count": 1}
2019-05-19T10:39:54.981-0700    INFO    operator.certificate-publisher-controller       certificate-publisher/controller.go:174 Reconciling     {"request": "openshift-ingress-operator/default"}
2019-05-19T10:39:54.981-0700    INFO    operator.controller     controller/controller.go:100    reconciling     {"request": "openshift-ingress-operator/default"}
2019-05-19T10:39:55.063-0700    DEBUG   operator.init.controller-runtime.controller     controller/controller.go:236    Successfully Reconciled {"controller": "certificate-publisher-controller", "request": "openshift-ingress-operator/default"}
2019-05-19T10:39:55.962-0700    DEBUG   operator.init.controller-runtime.controller     controller/controller.go:236    Successfully Reconciled {"controller": "certificate-controller", "request": "openshift-ingress-operator/default"}
2019-05-19T10:40:02.678-0700    ERROR   operator.init.controller-runtime.controller     controller/controller.go:217    Reconciler error        {"controller": "operator-controller", "request": "openshift-ingress-operator/default", "error": "failed to ensure ingresscontroller: [failed to ensure DNS for default: no load balancer is assigned to service openshift-ingress/router-default, failed to integrate metrics with openshift-monitoring for ingresscontroller default: failed to ensure servicemonitor for default: no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\"]", "errorCauses": [{"error": "failed to ensure ingresscontroller: [failed to ensure DNS for default: no load balancer is assigned to service openshift-ingress/router-default, failed to integrate metrics with openshift-monitoring for ingresscontroller default: failed to ensure servicemonitor for default: no matches for kind \"ServiceMonitor\" in version \"monitoring.coreos.com/v1\"]"}]}
2019-05-19T10:40:03.679-0700    INFO    operator.controller     controller/controller.go:100    reconciling     {"request": "openshift-ingress-operator/default"}
```

**Comment by serbrech** [comment]
rebased on master

File: pkg/operator/controller/controller_lb.go [hunk 3]
```diff
@@ -127,7 +127,7 @@ func (r *reconciler) finalizeLoadBalancerService(ci *operatorv1.IngressControlle
 	// an annotation on the ingresscontroller.
 	ingress := service.Status.LoadBalancer.Ingress
 	if len(ingress) > 0 && len(ingress[0].Hostname) > 0 {
-		records, err := desiredDNSRecords(ci, ingress[0].Hostname, dnsConfig)
+		records, err := desiredDNSAliasRecords(ci, ingress[0].Hostname, dnsConfig)
```
Comment (ironcladlou) [inline pkg/operator/controller/controller_lb.go]: Looks like this needs a similar treatment for `.IP` handling to avoid leaking A records

**Comment by ironcladlou** [comment]
I have some additional necessary changes in https://github.com/serbrech/cluster-ingress-operator/pull/1 which will need incorporated here.

**Comment by ironcladlou** [comment]
/approve

Will lift the hold once https://github.com/serbrech/cluster-ingress-operator/pull/1 merges.

**Comment by ironcladlou** [comment]
@serbrech many, many thanks for this one. I'm going to get it merged so we can iterate more quickly going forward.

/lgtm

**Pull request merged**